File: clients/go/model/game.go
package model

import "fmt"
import "io"
import . "ai_cup_22/stream"

// Current game's state
type Game struct {
    // Your player's id
    MyId int32
    // List of players (teams)
    Players []Player
    // Current tick
    CurrentTick int32
    // List of units visible by your team
    Units []Unit
    // List of loot visible by your team
    Loot []Loot
    // List of projectiles visible by your team
    Projectiles []Projectile
    // Current state of game zone
    Zone Zone
    // List of sounds heard by your team during last tick
    Sounds []Sound
}

func NewGame(myId int32, players []Player, currentTick int32, units []Unit, loot []Loot, projectiles []Projectile, zone Zone, sounds []Sound) Game {
    return Game {
        MyId: myId,
        Players: players,
        CurrentTick: currentTick,
        Units: units,
        Loot: loot,
        Projectiles: projectiles,
        Zone: zone,
        Sounds: sounds,
    }
}

// Read Game from reader
func ReadGame(reader io.Reader) Game {
    var myId int32
    myId = ReadInt32(reader)
    var players []Player
    players = make([]Player, ReadInt32(reader))
    for playersIndex := range players {
        var playersElement Player
        playersElement = ReadPlayer(reader)
        players[playersIndex] = playersElement
    }
    var currentTick int32
    currentTick = ReadInt32(reader)
    var units []Unit
    units = make([]Unit, ReadInt32(reader))
    for unitsIndex := range units {
        var unitsElement Unit
        unitsElement = ReadUnit(reader)
        units[unitsIndex] = unitsElement
    }
    var loot []Loot
    loot = make([]Loot, ReadInt32(reader))
    for lootIndex := range loot {
        var lootElement Loot
        lootElement = ReadLoot(reader)
        loot[lootIndex] = lootElement
    }
    var projectiles []Projectile
    projectiles = make([]Projectile, ReadInt32(reader))
    for projectilesIndex := range projectiles {
        var projectilesElement Projectile
        projectilesElement = ReadProjectile(reader)
        projectiles[projectilesIndex] = projectilesElement
    }
    var zone Zone
    zone = ReadZone(reader)
    var sounds []Sound
    sounds = make([]Sound, ReadInt32(reader))
    for soundsIndex := range sounds {
        var soundsElement Sound
        soundsElement = ReadSound(reader)
        sounds[soundsIndex] = soundsElement
    }
    return Game {
        MyId: myId,
        Players: players,
        CurrentTick: currentTick,
        Units: units,
        Loot: loot,
        Projectiles: projectiles,
        Zone: zone,
        Sounds: sounds,
    }
}

// Write Game to writer
func (game Game) Write(writer io.Writer) {
    myId := game.MyId
    WriteInt32(writer, myId)
    players := game.Players
    WriteInt32(writer, int32(len(players)))
    for _, playersElement := range players {
        playersElement.Write(writer)
    }
    currentTick := game.CurrentTick
    WriteInt32(writer, currentTick)
    units := game.Units
    WriteInt32(writer, int32(len(units)))
    for _, unitsElement := range units {
        unitsElement.Write(writer)
    }
    loot := game.Loot
    WriteInt32(writer, int32(len(loot)))
    for _, lootElement := range loot {
        lootElement.Write(writer)
    }
    projectiles := game.Projectiles
    WriteInt32(writer, int32(len(projectiles)))
    for _, projectilesElement := range projectiles {
        projectilesElement.Write(writer)
    }
    zone := game.Zone
    zone.Write(writer)
    sounds := game.Sounds
    WriteInt32(writer, int32(len(sounds)))
    for _, soundsElement := range sounds {
        soundsElement.Write(writer)
    }
}

// Get string representation of Game
func (game Game) String() string {
    stringResult := "{ "
    stringResult += "MyId: "
    myId := game.MyId
    stringResult += fmt.Sprint(myId)
    stringResult += ", "
    stringResult += "Players: "
    players := game.Players
    stringResult += "[ "
    for playersIndex, playersElement := range players {
        if playersIndex != 0 {
            stringResult += ", "
        }
        stringResult += playersElement.String()
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "CurrentTick: "
    currentTick := game.CurrentTick
    stringResult += fmt.Sprint(currentTick)
    stringResult += ", "
    stringResult += "Units: "
    units := game.Units
    stringResult += "[ "
    for unitsIndex, unitsElement := range units {
        if unitsIndex != 0 {
            stringResult += ", "
        }
        stringResult += unitsElement.String()
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "Loot: "
    loot := game.Loot
    stringResult += "[ "
    for lootIndex, lootElement := range loot {
        if lootIndex != 0 {
            stringResult += ", "
        }
        stringResult += lootElement.String()
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "Projectiles: "
    projectiles := game.Projectiles
    stringResult += "[ "
    for projectilesIndex, projectilesElement := range projectiles {
        if projectilesIndex != 0 {
            stringResult += ", "
        }
        stringResult += projectilesElement.String()
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "Zone: "
    zone := game.Zone
    stringResult += zone.String()
    stringResult += ", "
    stringResult += "Sounds: "
    sounds := game.Sounds
    stringResult += "[ "
    for soundsIndex, soundsElement := range sounds {
        if soundsIndex != 0 {
            stringResult += ", "
        }
        stringResult += soundsElement.String()
    }
    stringResult += " ]"
    stringResult += " }"
    return stringResult
}

File: clients/swift/Sources/AiCup22/Model/Game.swift
/// Current game's state
public struct Game {
    /// Your player's id
    let myId: Int32

    /// List of players (teams)
    let players: [Player]

    /// Current tick
    let currentTick: Int32

    /// List of units visible by your team
    let units: [Unit]

    /// List of loot visible by your team
    let loot: [Loot]

    /// List of projectiles visible by your team
    let projectiles: [Projectile]

    /// Current state of game zone
    let zone: Zone

    /// List of sounds heard by your team during last tick
    let sounds: [Sound]

    /// Read Game from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Game {
        var myId: Int32
        myId = stream.readInt32()
        var players: [Player]
        let playersSize = stream.readInt32()
        players = (0..<playersSize).map{ _ in
            var playersSize: Player
            playersSize = Player.readFrom(stream)
            return playersSize
        }
        var currentTick: Int32
        currentTick = stream.readInt32()
        var units: [Unit]
        let unitsSize = stream.readInt32()
        units = (0..<unitsSize).map{ _ in
            var unitsSize: Unit
            unitsSize = Unit.readFrom(stream)
            return unitsSize
        }
        var loot: [Loot]
        let lootSize = stream.readInt32()
        loot = (0..<lootSize).map{ _ in
            var lootSize: Loot
            lootSize = Loot.readFrom(stream)
            return lootSize
        }
        var projectiles: [Projectile]
        let projectilesSize = stream.readInt32()
        projectiles = (0..<projectilesSize).map{ _ in
            var projectilesSize: Projectile
            projectilesSize = Projectile.readFrom(stream)
            return projectilesSize
        }
        var zone: Zone
        zone = Zone.readFrom(stream)
        var sounds: [Sound]
        let soundsSize = stream.readInt32()
        sounds = (0..<soundsSize).map{ _ in
            var soundsSize: Sound
            soundsSize = Sound.readFrom(stream)
            return soundsSize
        }
        return Game(myId: myId, players: players, currentTick: currentTick, units: units, loot: loot, projectiles: projectiles, zone: zone, sounds: sounds)
    }

    /// Write Game to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(myId)
        stream.writeInt32(Int32(players.count))
        for playersElement in players {
            playersElement.writeTo(stream)
        }
        stream.writeInt32(currentTick)
        stream.writeInt32(Int32(units.count))
        for unitsElement in units {
            unitsElement.writeTo(stream)
        }
        stream.writeInt32(Int32(loot.count))
        for lootElement in loot {
            lootElement.writeTo(stream)
        }
        stream.writeInt32(Int32(projectiles.count))
        for projectilesElement in projectiles {
            projectilesElement.writeTo(stream)
        }
        zone.writeTo(stream)
        stream.writeInt32(Int32(sounds.count))
        for soundsElement in sounds {
            soundsElement.writeTo(stream)
        }
    }
}